Source/Integrations/KingfisherIntegration.swift

```diff
@@ -0,0 +1,75 @@
+//
+//  KingfisherIntegration.swift
+//  AXPhotoViewer
+//
+//  Created by Matt Lamont on 11/08/17.
+//  Copyright © 2017 Matt Lamont. All rights reserved.
+//
+
+import Kingfisher
+
+class KingfisherIntegration: NSObject, NetworkIntegrationProtocol {
+
+    weak public var delegate: NetworkIntegrationDelegate?
+
+    fileprivate var downloadTasks = NSMapTable<PhotoProtocol, RetrieveImageTask>(keyOptions: .strongMemory, valueOptions: .strongMemory)
+
+    public func loadPhoto(_ photo: PhotoProtocol) {
+        if photo.imageData != nil || photo.image != nil {
+            delegate?.networkIntegration(self, loadDidFinishWith: photo)
+        }
+
+        guard let url = photo.url else {
+            return
+        }
+
+        let progress: DownloadProgressBlock = { [weak self] (receivedSize, totalSize) in
+            guard let uSelf = self else {
+                return
+            }
+
+            uSelf.delegate?.networkIntegration?(uSelf, didUpdateLoadingProgress: CGFloat(totalSize / receivedSize), for: photo)
+        }
+
+        let completion: CompletionHandler = { [weak self] (image, error, cacheType, imageURL) in
+            guard let uSelf = self else {
+                return
+            }
+
+            self?.downloadTasks.removeObject(forKey: photo)
+
+            if let error = error {
+                uSelf.delegate?.networkIntegration(uSelf, loadDidFailWith: error, for: photo)
+            } else {
+                if let image = image {
```

Comment:
There should be a codepath in this if statement that allows `imageData` to be set as well (for gifs) if it exists - I believe Kingfisher exposes this property on the `Image` object. Something like `image.kf.gifRepresentation()`